### PR TITLE
feat(api): use conditional requests and fetch all inbox notifications

### PR DESCRIPTION
**Pull request by setchy**
Closes https://github.com/gitify-app/gitify/issues/1124
Closes https://github.com/gitify-app/gitify/issues/356

Update our API Request With Auth function to:
- support paginated endpoints and unpaginating response data
- conditionally set `Cache-Control` request header depending on the api url
  - for `/notifications` we do not want to cache request/response (current behavior)
  - for all other api requests, we use the default browser caching based on the GitHub API response (60 seconds)
  

Appreciate some hands to test this branch locally before merging 🙏

### CODE REVIEW

File: src/utils/api/utils.ts [hunk 1]
```diff
@@ -22,3 +23,9 @@ export function getGitHubGraphQLUrl(hostname: Hostname): URL {
 
   return url;
 }
+
+export function parseNextUrl(response: AxiosResponse): string | null {
```
Comment (bmulholland) [inline src/utils/api/utils.ts:27]: Parse for what? :)

Comment (bmulholland) [reply src/utils/api/utils.ts:27]: Should probably instead be named after what it's parsing out and returning.

**Review by bmulholland — APPROVED** [review]

File: src/utils/api/utils.ts (hunk 1 above)
Comment (setchy) [reply src/utils/api/utils.ts:27]: `getNextURLFromLinkHeader` ?

**Comment by setchy** [comment]
@bmulholland @afonsojramos - thoughts on whether we need to place a ceiling on the total number of notifications we unpaginate, or leave as-is (fetch all pages)?  

I've tested with ~650 notifications and it seems to be fine.  Maybe we add a ceiling only if there are reports of rate-limiting?

**Comment by setchy** [comment]
Moving back to draft as the tests require some rework

**Comment by bmulholland** [comment]
> thoughts on whether we need to place a ceiling on the total number of notifications we unpaginate, or leave as-is (fetch all pages)?

Maybe just leave it, but add a way to report to the user "you have too many notifications, it hits rate limits, you're on your own"? :)

**Comment by setchy** [comment]
Finally ready for this to be considered for merging :)

**Comment by bmulholland** [comment]
Will test locally. To start, I see this in console:
```
13:50:17.163 › Error occurred while fetching details for PullRequest notification: feat(lorem): add character length parameter to lorem.text TypeError: Cannot read properties of null (reading 'login')
```

**Comment by setchy** [comment]
> Will test locally. To start, I see this in console:
> 
> ```
> 13:50:17.163 › Error occurred while fetching details for PullRequest notification: feat(lorem): add character length parameter to lorem.text TypeError: Cannot read properties of null (reading 'login')
> ```

🤔

**Comment by setchy** [comment]
I've been running it locally for the past week and haven't had any console log entries like that

**Comment by setchy** [comment]
Added a new `Notification Setting` to control fetching **all** unread inbox notifications, just in case

Screenshot 2024-09-09 at 2 36 09 PM

**Comment by bmulholland** [comment]
@setchy the console message may well be unrelated, unfortunately I only had that PR in my notifications exactly while testing the branch. I think you're right, though: the PR author deleted their account.

**Comment by bmulholland** [comment]
FYI, I've been using this branch for the last day and it seems to be working well for me :)

**Comment by setchy** [comment]
> FYI, I've been using this branch for the last day and it seems to be working well for me :)

Likewise - do we feel comfortable enough to merge this change?  Itching to ship the next release

File: src/components/settings/NotificationSettings.tsx [hunk 2]
```diff
@@ -25,6 +25,26 @@ export const NotificationSettings: FC = () => {
...
+          updateSetting('fetchAllNotifications', evt.target.checked)
+        }
+        tooltip={
+          <div>
+            <div className="pb-3">
+              When <em>checked</em>, Gitify will fetch <strong>all</strong>{' '}
+              notifications from your inbox.
+            </div>
+            <div>
+              When <em>unhecked</em>, Gitify will only fetch the first page of
```
Comment (afonsojramos) [inline src/components/settings/NotificationSettings.tsx:42]:
```suggestion
              When <em>unchecked</em>, Gitify will only fetch the first page of
```

File: src/components/settings/NotificationSettings.tsx [hunk 3]
```diff
@@ -25,6 +25,26 @@ export const NotificationSettings: FC = () => {
...
+              notifications (max 50 records pet GitHub account)
```
Comment (afonsojramos) [inline src/components/settings/NotificationSettings.tsx:43]:
```suggestion
              notifications (max 50 records per GitHub account)
```

**Review by afonsojramos — APPROVED** [review]

**Pull request merged**